### PR TITLE
Image augmentation pipeline

### LINKED ISSUE #8: Image augmentation pipeline
**Section 1: Design details**

We want to have an image augmentation pipeline that randomly transforms the images. PR #9 

**Relevant background**

DL models are known to become more robust and overfit less when augmentations are used during training. Thus we want to have an augmentation pipeline that applies spatial and intensity augmentations onto images and solely spatial augmentations on the masks.

**Design overview**

- `get_augmentation_pipeline(params)` function that takes in a parameter dict and returns an object of class `imagaug.augmenters.Sequential`.
- `augment_images(images, masks, augmentation_pipeline)` function that takes in the augmentation pipeline, input images and masks and returns augmented images and masks

**Design list/flowchart**

- `get_augmentation_pipeline(params)` produces pipeline object
- `augment_images(images, masks, augmentation_pipeline)` applies augmentation pipeline onto our data. For images bilinear interpolation is used, for masks nearest-neighbor interpolation so that their unique values don't change.

**Required inputs**

- `params = {
        # flip
        "flip_prob": 1.0,
        # affine
        "affine_prob": 1.0, "scale_min": 0.5, "scale_max": 1.5, "shear_angle": 10,
        # elastic
        "elastic_prob": 1.0, "elastic_alpha": [0, 5.0], "elastic_sigma": 0.5,
        # rotate
        "rotate_count": [0, 3],
        # gaussian noise
        "gaussian_noise_prob": 1.0, "gaussian_noise_min": 0.1, "gaussian_noise_max": 0.5,
        # gaussian blur
        "gaussian_blur_prob": 0.0, "gaussian_blur_min": 0.1, "gaussian_blur_max": 0.5,
        # contrast aug
        "contrast_prob": 0.0, "contrast_min": 0.1, "contrast_max": 2.0,
    }`
- images (np.array, dtype=float32) B x H x W x C
- masks (np.array, dtype=uint8/16) B x H x W x C

**Output files**

Intermediate output: augmentation pipeline object of class `imagaug.augmenters.Sequential`
Final output: augmented images (np.array, dtype=float32) B x H x W x C and masks (np.array, dtype=uint8/16) B x H x W x C

**Section 2: Implementation details**

**Control flow**

Images need to be interpolated with bilinear interpolation, masks need to be interpolated with nearest-neighbor interpolation, otherwise the they will contain label values that weren't there before.

**Milestones and timeline**

Ready for review.

### PR DESCRIPTION
**Pull request by JLrumberger**
**What is the purpose of this PR?**

Adding an image augmentation pipeline as closes #8 .

**How did you implement your changes**

With `imgaug` library. We have one function that constructs the image augmentation pipeline and another function that takes the pipeline and the data and transforms it.

**Remaining issues**

None

### CODE REVIEW

File: cell_classification/augmentation_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    Returns:
+        np.array:
+            The augmented images
+        np.array:
+            The augmented masks
+    """
+    masks = [SegmentationMapsOnImage(mask, shape=images.shape[1:-1]) for mask in masks]
+    batch = Batch(images=images, segmentation_maps=masks)
+    batch = augmentation_pipeline.augment_batch_(batch)
+    augmented_masks = [np.squeeze(mask.arr) for mask in batch.segmentation_maps_aug]
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline.py]: Good to add a comment about why the shape is changing

File: cell_classification/augmentation_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    return np.stack(batch.images_aug, 0), np.squeeze(np.stack(augmented_masks, 0))
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline.py]: How come we squeeze the `augmented_masks` twice?

File: cell_classification/augmentation_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    Args:
+        params (dict):
+            The parameters for the augmentation
+    Returns:
+        imgaug.augmenters.meta.Augmenter:
+            The augmentation pipeline
+    """
+    augmentation_pipeline = iaa.Sequential(
+        [
+            iaa.Fliplr(params["flip_prob"]),
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline.py:49]: Brief comments defining the transformations

File: cell_classification/augmentation_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
+import imgaug as ia
+import imgaug.augmenters as iaa
+from imgaug.augmentables.segmaps import SegmentationMapsOnImage
+from imgaug.augmentables.batches import Batch
+import numpy as np
+
+
+def augment_images(images, masks, augmentation_pipeline):
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline.py:8]: We'll need to make sure the tf.record follows this same format of separate stacks for images and masks

File: cell_classification/augmentation_pipeline_test.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    # check if right types and shapes are returned
+    assert type(augmented_images) == np.ndarray
+    assert type(augmented_masks) == np.ndarray
+    assert augmented_images.dtype == np.float32
+    assert augmented_masks.dtype == np.int32
+    assert augmented_images.shape == images.shape
+    assert augmented_masks.shape == masks.shape
+
+    # check if images are augmented
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline_test.py:55]: Perfect level of commenting here for the tests

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks great! Just a few areas where some more comments would be helpful

File: cell_classification/augmentation_pipeline.py (hunk 2 above)
Comment (JLrumberger) [reply cell_classification/augmentation_pipeline.py]: This is an error. Should only be squeezed once. The problem is the following: `augmentation_pipeline.augment_batch_(batch)` adds a channel dimension to the mask if there's none, to ensure that input and output dim are the same, I think I should check if a channel dim got added, and if this happened it should be squeezed. Because as it is now, it would also squeeze the batch dim for batch size 1 which is unwanted.

File: cell_classification/augmentation_pipeline.py (hunk 1 above)
Comment (JLrumberger) [reply cell_classification/augmentation_pipeline.py]: It doesn't change. I changed this to `shape=mask.shape` to make it more clear.

File: cell_classification/augmentation_pipeline.py (hunk 4 above)
Comment (JLrumberger) [reply cell_classification/augmentation_pipeline.py:8]: Yep, this goes into the `prepare_example` function. We should do it anyways, because the actual mibi image will be the only thing that is stored as float, for the rest we'll use the minimal data type (i.e. uint8 for binary mask, uint16 instance_mask) to reduce i/o as much as possible.

File: cell_classification/augmentation_pipeline.py (hunk 2 above)
Comment (ngreenwald) [reply cell_classification/augmentation_pipeline.py]: Won't the masks always have a channels dim?

We should loop over a couple different shapes for the testing data. Right now it's `[3, 100, 100, 2]`. We should probably check the case where both the first and/or last index is 1.

Comment (JLrumberger) [reply cell_classification/augmentation_pipeline.py]: Changed it so that it works for masks with or without channel dim and with different number of channels.

File: cell_classification/augmentation_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        np.array:
+            The augmented masks
+    """
+    masks_ = [SegmentationMapsOnImage(mask, shape=mask.shape) for mask in masks]
+    batch = Batch(images=images, segmentation_maps=masks_)
+    batch = augmentation_pipeline.augment_batch_(batch)
+    augmented_masks = [mask.arr for mask in batch.segmentation_maps_aug]
+    augmented_masks = np.stack(augmented_masks, 0)
+
+    # remove additional channel from single channel masks
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline.py:30]: Nice, this is more robust and easier to understand

File: cell_classification/augmentation_pipeline_test.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    # check if masks are still binary with the right labels
+    assert list(np.unique(augmented_masks)) == [0, 1, 2]
+
+    # check if images and masks where augmented with the same spatial augmentations approx.
+    assert np.abs(augmented_images[augmented_masks == 0].mean() - images[masks == 0].mean()) < 1
+    assert np.abs(augmented_images[augmented_masks == 1].mean() - images[masks == 1].mean()) < 1
+    assert np.abs(augmented_images[augmented_masks == 2].mean() - images[masks == 2].mean()) < 5
+
+    # check control flow for different number of samples per batch and channels
+    for b in [1, 2, 3]:
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline_test.py]: This is great. Instead of a for loop inside the test, you can use pytest to automatically pass the correct variable. Then you can get rid of this separate section and test the batch sizes across the whole function
https://github.com/angelolab/toffy/blob/acd6df4d4b71484d2aae5629094cc5ebb7cd7bd6/toffy/tiling_utils_test.py#L405

```
parametrize = pytest.mark.parametrize

@parametrize('batch_num', [1, 2, 3])
@parametrize('chan_num', [1,2,3])
def test_augment_images(batch_num, chan_num)
...

masks = np.zeros([batch_num 100, 100, chan_num], dtype=np.int32)
```

File: cell_classification/augmentation_pipeline_test.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        for c in [1, 2, 3]:
+            images = np.zeros([b, 100, 100, c], dtype=np.float32)
+            masks = np.zeros([b, 100, 100, c], dtype=np.int32)
+            # if c == 1:
+            #     masks = np.squeeze(masks, axis=-1)
+            _, augmented_masks = augment_images(images, masks, augmentation_pipeline)
+            assert augmented_masks.shape == masks.shape
+
+    # check control flow for no channel dimensions for the masks
+    masks = np.zeros([b, 100, 100], dtype=np.int32)
```
Comment (ngreenwald) [inline cell_classification/augmentation_pipeline_test.py]: You can either keep this as a separate test case, or do something like this at the top:

`mask = np.zeros((batch_num, 100, 100)) if chan_num == 0 else mask = np.zeros((batch_num, 100, 100, chan_num)`

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks awesome, here's what I was talking about for letting pytest handle the for loop. There's often scenarios where you want to loop over different parameters to test, this is really useful.

**Comment by ngreenwald** [comment]
Also, make sure to update your branch. Can either do it locally or from github. 
image

File: cell_classification/augmentation_pipeline_test.py (hunk 7 above)
Comment (JLrumberger) [reply cell_classification/augmentation_pipeline_test.py]: Thanks for the suggestion, looks way better with the decorator!

File: cell_classification/augmentation_pipeline_test.py (hunk 8 above)
Comment (JLrumberger) [reply cell_classification/augmentation_pipeline_test.py]: I'll just keep it as is, because it's easier to read I think.
